src/node.rs

```diff
@@ -0,0 +1,153 @@
+use crate::helper::Identifier;
+use crate::kbucket::{Bucket, KbucketTable, TableRecord};
+use std::collections::HashMap;
+use std::net::{Ipv4Addr, SocketAddrV4};
+use tokio::io;
+use tokio::net::UdpSocket;
+
+// The main Kademlia client struct.
+// Provides user-level API for performing querie and interacting with the underlying service.
+#[derive(Debug, PartialEq)]
+pub struct Node {
+    pub node_id: Identifier,
+    pub table: KbucketTable,
+    pub store: HashMap<Vec<u8>, Vec<u8>>, // Same storage as Portal network to store samples
+}
+
+impl Node {
+    pub fn new(node_id: Identifier, table_record: TableRecord) -> Self {
+        Self {
+            node_id,
+            table: KbucketTable::new(node_id, table_record),
+            store: Default::default(),
+        }
+    }
+
+    // Protocol's RPCs:
+    // ---------------------------------------------------------------------------------------------------
+    /// "The most important procedure a Kademlia participant must perform is to locate
+    /// the k closest nodes to some given node ID"
+    ///     - Kademlia Paper
+    ///
+    /// Recieves an id request and returns node information on nodes within
+    /// *its closest bucket* (instead of k-closest nodes) to that id.
+    pub fn find_node(&mut self, id: &Identifier) -> Vec<TableRecord> {
+        let bucket_index = self.table.xor_bucket_index(id);
+        self.table.buckets[bucket_index]
+            .map
+            .values()
+            .cloned()
+            .collect()
```

Comment:
nice! you could imagine moving some of these details to the `Table`

so the API is more like
```rust
self.table.get_bucket_for(id) // -> &Bucket
  .clone() // if you want to just clone all of the values
```

and all of the logic you have here is encapsulated in the `get_bucket_for` method